plugins/outputs/cloudwatch/cloudwatch.go

```diff
@@ -279,45 +277,86 @@ func (c *CloudWatch) timeToPublish(b *MetricDatumBatch) bool {
+// it flushed immediately. Now it cuts the remaining interval in half,
+// and the next interval in half.
+// If the buffer continues to stay full then it will keep cutting down.
+// This avoids bursting the backend and maintains some amount of jitter.
 func (c *CloudWatch) publish() {
-	now := time.Now()
-	forceFlushInterval := c.ForceFlushInterval.Duration
-	publishJitter := publishJitter(forceFlushInterval)
-	log.Printf("I! cloudwatch: publish with ForceFlushInterval: %v, Publish Jitter: %v", forceFlushInterval, publishJitter)
-	time.Sleep(now.Truncate(forceFlushInterval).Add(publishJitter).Sub(now))
-	c.pushTicker = time.NewTicker(c.ForceFlushInterval.Duration)
-	defer c.pushTicker.Stop()
-	shouldPublish := false
+	currentInterval := c.ForceFlushInterval.Duration
+	nextMs := getFirstPushMs(currentInterval)
+	// Only allow shortening interval once per push.
+	// This avoids bursting many pushes all at once, but still allows the
+	// push frequency to increase.
+	bufferFullOccurred := false
+
 	for {
+		shouldPublish := false
 		select {
 		case <-c.shutdownChan:
-			log.Printf("D! CloudWatch: publish routine receives the shutdown signal, exiting.")
+			log.Printf("D! cloudwatch: publish routine receives the shutdown signal, exiting.")
 			return
-		case <-c.pushTicker.C:
-			shouldPublish = true
 		case <-c.aggregatorShutdownChan:
 			shouldPublish = true
-		case <-c.metricDatumBatchFull():
-			shouldPublish = true
 		default:
-			shouldPublish = false
 		}
+
+		nowMs := time.Now().UnixMilli()
+
+		if c.metricDatumBatchFull() {
+			if !bufferFullOccurred {
+				// Set to false so this only happens once per push.
+				bufferFullOccurred = true
+				// Keep interval above 1 second.
+				if currentInterval.Seconds() >=  2 {
+					currentInterval /= 2
+					nextMs = nowMs + ((nextMs - nowMs) / 2)
+					log.Printf("I! cloudwatch: cut interval to %v", currentInterval)
+				}
+			}
+		}
+
+		// Check if the interval has elapsed.
+		if nowMs >= nextMs {
+			shouldPublish = true
+			nextMs = nowMs + currentInterval.Milliseconds()
+			// Restore interval if buffer did not fill up during this interval.
+			if !bufferFullOccurred {
+				log.Printf("I! cloudwatch: reset interval")
+				currentInterval = c.ForceFlushInterval.Duration
+			}
+		}
+
 		if shouldPublish {
 			c.pushMetricDatumBatch()
-		} else {
-			time.Sleep(time.Second)
+			bufferFullOccurred = false
 		}
+		// Always sleep, so the fastest we push is every 1 second.
+		time.Sleep(time.Second)
 	}
 }
 
-func (c *CloudWatch) metricDatumBatchFull() chan bool {
-	if len(c.datumBatchChan) >= datumBatchChanBufferSize {
-		if len(c.datumBatchFullChan) == 0 {
-			c.datumBatchFullChan <- true
-		}
-		return c.datumBatchFullChan
-	}
-	return nil
+// metricDatumBatchFull returns true if the channel/buffer of batches if full.
+func (c *CloudWatch) metricDatumBatchFull() bool {
+	return len(c.datumBatchChan) >= datumBatchChanBufferSize
```

Comment:
This was not a necessary change.
It is simpler and clearer to just check this condition by calling a function instead of trying to recv from a channel.